title: "de-Identifier"
shorttitle: "deIdentify
package: sesame
output: rmarkdown::html_vignette
fig_width: 8
fig_height: 6
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{"0. de-Identifier"}
  %\VignetteEncoding{UTF-8} 
---

#' This requires the R package seSAMe
#' De-Identify IDATs by masking SNP probe intensity mean by zero

```{r message=FALSE, warning=FALSE, include=FALSE}
library(sesame)
```

# sesameData

#' Find high-quality DNA methylation data on more than 10,000 human samples with the HM450 platform

```{r, eval=FALSE}
dest_dir = tempdir()
res_grn = sesameDataDownload("3999492009_R01C01_Grn.idat", dest_dir=dest_dir)
res_red = sesameDataDownload("3999492009_R01C01_Red.idat", dest_dir=dest_dir)
```
                                                   
# deIdentify

#' Modifies IDAT files in place so they can be released to public domain without privacy leak
#' As a consequence, the allele frequency will be 0.5

```{r, eval=FALSE}

deIdentify(res_grn$dest_file, sprintf("%s/deidentified_Grn.idat", dest_dir))
deIdentify(res_red$dest_file, sprintf("%s/deidentified_Red.idat", dest_dir))

betas1 = getBetas(readIDATpair(sprintf("%s/3999492009_R01C01", dest_dir)))
betas2 = getBetas(readIDATpair(sprintf("%s/deidentified", dest_dir)))

head(betas1[grep('rs',names(betas1))]) # before deIdentify, the rs values will all be different 
head(betas2[grep('rs',names(betas2))]) # after deIdentify, all rs values will be 0.5
```

# reIdentify

#' Re-identify IDATs by restoring scrambled SNP intensities

```{r, eval=FALSE}
reIdentify <- function(path, out_path=NULL, snps=NULL, mft=NULL) {

    res <- suppressWarnings(illuminaio::readIDAT(path))
    platform <- inferPlatform(res)

    if(is.null(out_path)) {
        pfx <- sub('.idat(.gz)?$','', path)
        if(grepl('_Grn$', pfx)) {
            out_path <- paste0(sub('_Grn$','',pfx), '_reid_Grn.idat')
        } else if (grepl('_Red$', pfx)) {
            out_path <- paste0(sub('_Red$','',pfx), '_reid_Red.idat')
        }
 ```
    
# formactVCF

#' Convert SNP from Infinium array to VCF file

```{r, eval=FALSE}
formatVCF <- function(
    sset, vcf=NULL, refversion="hg19", annoS=NULL, annoI=NULL) {

    platform <- sset@platform

    if (is.null(annoS)) annoS <- sesameDataPullVariantAnno_SNP(platform)
    betas <- getBetas(sset)[names(annoS)]
    vafs <- ifelse(annoS$U == 'REF', betas, 1-betas)
    gts <- lapply(vafs, genotype)
    GT <- vapply(gts, function(g) g$GT, character(1))
    GS <- vapply(gts, function(g) g$GS, numeric(1))
    vcflines_snp <- cbind(as.character(GenomicRanges::seqnames(annoS)),
        as.character(end(annoS)),
        names(annoS),
        annoS$REF, annoS$ALT,
        GS, ifelse(GS>20,'PASS','FAIL'),
        sprintf("PVF=%1.3f;GT=%s;GS=%d", vafs, GT, GS))
```

