title: "de-Identifier"
shorttitle: "deIdentify
package: sesame
output: rmarkdown::html_vignette
fig_width: 8
fig_height: 6
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{"0. de-Identifier"}
  %\VignetteEncoding{UTF-8} 
---

Will need R package seSAMe
De-Identify IDATs by masking SNP probe intensity mean by zero
Modifies IDAT files in place so they can be released to public domain without privacy leak
As a consequence, the allele frequency will be 0.5

```{r message=FALSE, warning=FALSE, include=FALSE}
library(sesame)
```

# sesameData

Can find high-quality DNA methylation data on more than 10,000 human samples with the HM450 platform

```{r, eval=FALSE}
dest_dir = tempdir()
res_grn = sesameDataDownload("3999492009_R01C01_Grn.idat", dest_dir=dest_dir)
res_red = sesameDataDownload("3999492009_R01C01_Red.idat", dest_dir=dest_dir)
```
                                                   
# deIdentify

DNA methylation readouts from these arrays are usually expressed in Î² values

```{r, eval=FALSE}

deIdentify(res_grn$dest_file, sprintf("%s/deidentified_Grn.idat", dest_dir))
deIdentify(res_red$dest_file, sprintf("%s/deidentified_Red.idat", dest_dir))

betas1 = getBetas(readIDATpair(sprintf("%s/3999492009_R01C01", dest_dir)))
betas2 = getBetas(readIDATpair(sprintf("%s/deidentified", dest_dir)))

head(betas1[grep('rs',names(betas1))]) # before deIdentify, the rs values will all be different 
head(betas2[grep('rs',names(betas2))]) # after deIdentify, all rs values will be 0.5
```



