---
title: "Working with Non-human Array"
date: "`r BiocStyle::doc_date()`"
package: sesame
output: BiocStyle::html_document
fig_width: 6
fig_height: 5
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{2. Non-human Array}
  %\VignetteEncoding{UTF-8}
---

```{r message=FALSE, warning=FALSE, include=FALSE}
library(wheatmap)
library(dplyr)
options(rmarkdown.html_vignette.check_title = FALSE)
```

SeSAMe provides extensive native support for the Illumina mouse array (referred
to as the MM285 array) and the Horvath Mammal40 array (refered to as Mammal40
array). The MM285 array contains ~285,000 probes covering over 20 design
categories including gene promoters, enhancers, CpGs in synteny to human EPIC
array as well as other biology. This documents describe the procedure to
process the MM285 and the Mammal40 array. To begin, let's retrieve mouse
annotation data from ExperimentHub.  This only needs to be done once per sesame
installation.

```{r nh1, echo=FALSE, message=FALSE}
library(sesame)
sesameDataCache()
```

# The openSesame pipeline

The `openSesame` pipeline fully supports the MM285 and the Mammal40 arrays.
Let's download an example mouse array IDAT and call `openSesame` function.

```{r nh2, message=FALSE, eval=FALSE}
tmp = tempdir()
sesameAnno_download("Test/204637490002_R05C01_Grn.idat", dest_dir=tmp)
sesameAnno_download("Test/204637490002_R05C01_Red.idat", dest_dir=tmp)
head(openSesame(sprintf("%s/Test/204637490002_R05C01", tmp), prep="TQCDPB"))
```

Here is an example of `openSesame` processing on a Mammal40 array.

```{r nh3, message=FALSE, eval=FALSE}
tmp = tempdir()
sesameAnno_download("Test/GSM4411982_Grn.idat.gz", dest_dir=tmp)
sesameAnno_download("Test/GSM4411982_Red.idat.gz", dest_dir=tmp)
head(openSesame(sprintf("%s/Test/GSM4411982", tmp), prep="HSCDPM"))
```

The following sections provide a more detailed discussion of subsequent
inference you can do with these array data.

# MM285 array

## Replicate Probes

As you notice the probe ID has the suffix which distinguishes replicate
probes. These replicate probes can be collapsed to the cg prefix by adding the
`collapseToPfx=TRUE` option in `openSesame` and `getBetas` function (if that
function is used explicitly).

```{r nh4, message=TRUE, eval=FALSE}
sdf = sesameDataGet('MM285.1.SigDF')
betas = openSesame(sdf, prep="TQCDPB", collapseToPfx=TRUE)
```

## Quality Mask

For the default C57BL/6J strain, one can use the `qualityMask` (Q) function
without performing species or strain inference.

```{r nh5, message=TRUE, eval=FALSE}
sdf = sesameDataGet('MM285.1.SigDF')
betas = openSesame(sdf, prep="QCDPB")
```

By default the repeat and suboptimally designed probes are masked by `NA`.
These suboptimally designed probes may take a new probe ID prefix ("uk") in
addition to the "cg"/"ch"/"rs" typically seen. These masking is done by the
`qualityMask` function (or `Q` in prep codes).

```{r nh6}
sum(is.na(getBetas(sdf)))
```

To override these probes, one can use the `resetMask` function (the `0` code in
`openSesame`) to remove the default masking. We recommend using it after
preprocessing function so that preprocessing can only use reliable probes. Note
that there are still remaining NAs due to detection p-values (from pOOBAH):

```{r nh7, message=TRUE}
sum(is.na(openSesame(sdf, prep="TQCD0PB")))
```

## Strain Analysis

one need to call the `inferStrain` function
before calling the standard `noob`, `dyeBiasNL`, etc. This can be done by
simply adding "T" before "CDPB" when calling `openSesame`.

```{r nh8}
head(openSesame(sdf, prep="TQCDPB"))
```

The `inferStrain` function is what was used under the hood. This function will
return a new SigDF with col and mask updated to reflect the change of
strain. Optionally, one can also specify `return.strain=TRUE`,
`return.pval=TRUE` or `return.probability = TRUE` to return the inferred strain,
the p-value or the probabilities of all 37 strain candidates.  Internally, the
function converts the beta values to variant allele frequencies. It should be
noted that since variant allele frequency is not always measured by the
M-allele of the probe. SeSAMe flips the beta values for some probes to
calculate variant allele frequency.

```{r nh9}
inferStrain(sdf, return.strain = TRUE) # return strain as a string
sdf1 = inferStrain(sdf)
sum(sdf$mask) # before strain inference
sum(sdf1$mask) # after strain inference
```

Let's visualize the probabilities of all candidate strains.

```{r nh10, fig.width=6, fig.height=4}
library(ggplot2)
p = inferStrain(sdf, return.probability = TRUE)
df = data.frame(strain=names(p), probs=p)
ggplot(data = df,  aes(x = strain, y = probs)) +
  geom_bar(stat = "identity", color="gray") +
  ggtitle("Strain Probabilities") +
  ylab("Probability") + xlab("") +
  scale_x_discrete(position = "top") +
  theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust=0),
    legend.position = "none")
```

One can also visualize the actual SNP data for strain inference by the
following. The target sample is plotted as a vertical bar on the right of the
signature heatmap.

```{r nh11, fig.width=6, fig.height=8, eval=FALSE}
compareMouseStrainReference(getBetas(sdf))
```

As expected, the variant configuration is consistent with the NOD strain.

## Non-mouse Species

```{r nh12}
sdf = sesameDataGet("MM285.1.SigDF") # an example SigDF
```

Infer species (obviously this is a mouse, but this is supposed to work on rat,
human etc). Currently this feature supports the human arrays, the mouse array
and the mammal array.

```{r nh13, message=FALSE, eval=FALSE}
inferSpecies(sdf, return.species = TRUE)
```

## Tissue Type

Let's load beta values from SeSAMeData
```{r nh14}
betas = sesameDataGet("MM285.10.tissue")$betas[,1:2]
```

Compare mouse array data with mouse tissue references. This will return a grid
object that contrasts the traget sample with pre-built mouse tissue reference.
```{r nh15, eval=FALSE}
compareMouseTissueReference(betas)
```

## Sex Analysis

Let's load beta values from sesameData
```{r nh16}
sdf = sesameDataGet("MM285.1.SigDF")
```

Sex inference can take either a `SigDF` or a $\beta$ value vector.

```{r nh17, message=FALSE}
inferSex(sdf)
```

## Epigenetic Clock

Let's load beta values from SeSAMeData
```{r nh18}
betas = sesameDataGet('MM285.10.tissue')$betas
```

The age of the mouse can be predicted using the `predictMouseAgeInMonth`
function. This looks for overlapping probes and estimates age using an aging
model built from 347 MM285 probes. The function returns a numeric output of age
in months. The model is most accurate with SeSAMe preprocessing.  Here's an
example.

```{r nh19}
predictMouseAgeInMonth(betas[,1])
```
This indicates thaat this mouse is approximately 1.41 months old.

## Human-Mouse Mixture

UNDER CONSTRUCTION

# Horvath Mammal40 array

```{r nh20}
inferSpecies(sesameDataGet("Mammal40.1.SigDF"), return.species = TRUE)
```

# Visualize $\beta$ Values

The track visualization is supported for these arrays.

```{r nh21, message=FALSE}
betas = sesameDataGet("MM285.10.tissue")$betas
visualizeGene("Igf2", betas)
```

# Session Info

```{r}
sessionInfo()
```