---
title: "Working with Non-human Array"
date: "`r BiocStyle::doc_date()`"
package: sesame
output: BiocStyle::html_document
fig_width: 6
fig_height: 6
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{2. Non-human Array}
  %\VignetteEncoding{UTF-8}
---

```{r message=FALSE, warning=FALSE, include=FALSE}
library(sesame)
library(wheatmap)
library(dplyr)
options(rmarkdown.html_vignette.check_title = FALSE)
```

# Mouse MM285 Array

To begin, we need to retrieve mouse annotation data from ExperimentHub.
This only needs to be done once per sesame installation.
```{r echo=FALSE, message=FALSE}
sesameDataCacheAll()
```

SeSAMe provides extensive native support for the Illumina mouse array
(referred to as the MM285 array).
The MM285 contains ~285,000 probes covering over 20 design categories including
gene promoters, enhancers, CpGs in synteny to human EPIC array as well as other
biology. This documents describe the procedure to process the MM285 array.

Let's download an example mouse array IDAT
```{r eval=FALSE}
tmp <- tempdir()
sesameData_getAnno("test/204637490002_R05C01_Grn.idat", dest_dir=tmp)
sesameData_getAnno("test/204637490002_R05C01_Red.idat", dest_dir=tmp)
pfx = sprintf("%s/test/204637490002_R05C01", tmp)
```

To load IDAT into `SigSet`, one needs the readIDATpair function,
```{r eval=FALSE}
sdf = readIDATpair(pfx)
```

## Preprocessing

The default openSesame pipeline works for the default C57BL/6J strain. The
function processes the IDATs to produce beta values.

```{r}
sdf <- sesameDataGet('MM285.1.SigDF')
head(getBetas(pOOBAH(noob(dyeBiasNL(sdf)))))
```

To work on a different strain, one can call the `inferStrain` function before
calling the standard `noob`, `dyeBiasNL`, etc. More detail about `inferStrain`
can be found below.

```{r}
head(getBetas(pOOBAH(noob(dyeBiasNL(inferStrain(sdf))))))
```

or to cg prefix only

```{r}
head(getBetas(pOOBAH(noob(dyeBiasNL(inferStrain(sdf)))), collapseToPfx=TRUE))
```

By default the repeat and suboptimally designed probes are masked by `NA`.
These suboptimally designed probes may take a new probe ID prefix ("uk") in
addition to the "cg"/"ch"/"rs" typically seen. These masking is turned on by
default when IDAT files are read

```{r}
sum(is.na(getBetas(sdf)))
```

To override these probes, one can use the `resetMask` function to remove the
default masking. We recommend using it after preprocessing function so that
preprocessing only use reliable probes. Note that there are still remaining NAs
due to detection p-values (from pOOBAH):

```{r}
sum(is.na(getBetas(pOOBAH(resetMask(noob(dyeBiasNL(sdf)))))))
```

## Strain/species inference

```{r}
sdf <- sesameDataGet("MM285.1.SigDF") # an example SigDF
```

Infer species (obviously this is a mouse, but this is supposed to work on rat,
human etc). Currently this feature supports both the mouse array and the mammal
array.
```{r}
inferSpecies(sdf, return.species = TRUE)
```

Infer strain information for mouse array. This will return a new SigDF with col
and mask updated to reflect the change of strain. Optionally, one can also
specify `return.strain=TRUE`, `return.pval=TRUE` or return.probability = TRUE`
to return the inferred strain, the p-value or the probabilities of all 37
strain candidates.  Internally, the function converts the beta values to
variant allele frequencies. It should be noted that since variant allele
frequency is not always measured by the M-allele of the probe. SeSAMe flips the
beta values for some probes to calculate variant allele frequency.

```{r}
inferStrain(sdf, return.strain = TRUE) # return strain as a string
sdf1 <- inferStrain(sdf)
sum(sdf$mask) # before strain inference
sum(sdf1$mask) # after strain inference
```

Let's visualize the probabilities of other strains.

```{r fig.width=6, fig.height=4}
library(ggplot2)
p <- inferStrain(sdf, return.probability = TRUE)
df <- data.frame(strain=names(p), probs=p)
ggplot(data = df,  aes(x = strain, y = probs)) +
  geom_bar(stat = "identity", color="gray") +
  ggtitle("Strain Probabilities") +
  ylab("Probability") + xlab("") +
  scale_x_discrete(position = "top") +
  theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust=0),
    legend.position = "none")
```

One can also visualize the actual SNP data for strain inference by the
following. The target sample is plotted as a vertical bar on the right of the
signature heatmap.

```{r fig.width=6, fig.height=8}
compareMouseStrainReference(getBetas(sdf))
```

As expected, the variant configuration is consistent with the NOD strain.

## Track View
```{r message=FALSE}
betas = sesameDataGet("MM285.10.tissue")$betas
visualizeGene("Igf2", betas = betas, platform="MM285", refversion = "mm10")
```

## Tissue Type Inference

Let's load beta values from SeSAMeData
```{r}
betas <- sesameDataGet("MM285.10.tissue")$betas[,1:2]
```

Compare mouse array data with mouse tissue references. This will return a grid
object that contrasts the traget sample with pre-built mouse tissue reference.
```{r}
compareMouseTissueReference(betas)
```

## Sex Inference

Let's load beta values from sesameData
```{r}
sdf = sesameDataGet("MM285.1.SigDF")
```

Sex inference can take both the raw signal in SigDF or beta value vector
```{r}
inferSex(sdf)
```

## Age Inference

Let's load beta values from SeSAMeData
```{r}
betas <- sesameDataGet('MM285.10.tissue')$betas
```

The age of the mouse can be predicted using the `predictMouseAgeInMonth`
function. This looks for overlapping probes and estimates age using an aging
model built from 347 MM285 probes. The function returns a numeric output of
age in months. The model is most accurate with SeSAMe preprocessing.
Here's an example.
```{r}
predictMouseAgeInMonth(betas[,1])
```
This indicates thaat this mouse is approximately 1.41 months old.

## Human-Mouse Mixture

UNDER CONSTRUCTION

# Horvath Mammal40 Array

SeSAMe supports Mammal 40 array the same way:

```{r}
tmp = tempdir()
sesameData_getAnno("test/GSM4411982_Grn.idat.gz", dest_dir=tmp)
sesameData_getAnno("test/GSM4411982_Red.idat.gz", dest_dir=tmp)
sdf = readIDATpair(sprintf("%s/test/GSM4411982", tmp))
```

Preprocess the sigset to produce beta values.
The standard `noob`, `dyeBiasCorrTypeINorm` works as expected:
```{r}
sdf_normalized = dyeBiasCorrTypeINorm(noob(pOOBAH(sdf)))
```
Retrieve beta values using the following commands
```{r}
betas = getBetas(sdf_normalized)
head(betas)
```

# Session Info

```{r}
sessionInfo()
```