<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>KnowYourCpG</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css">
<link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css" integrity="sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T" crossorigin="anonymous">
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-92943008-1', 'auto');
  ga('send', 'pageview');
</script>
<link rel="stylesheet" href="include/site.css" type="text/css" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
      .display.math{display: block; text-align: center; margin: 0.5rem auto;}
  </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<link rel="stylesheet" href="include/site.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><a href="http://bioconductor.org/packages/release/bioc/html/sesame.html" class="navbar-left"><img src="include/logo.png" height="50" width="50" vspace=5 hspace=10></a></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="sesame.html">
    <span class="fa fa-home"></span>
     
    Basics
  </a>
</li>
<li>
  <a href="QC.html">
    <span class="fa fa-copy"></span>
     
    Quality Control
  </a>
</li>
<li>
  <a href="nonhuman.html">
    <span class="fa fa-paw"></span>
     
    Non-human Array
  </a>
</li>
<li>
  <a href="modeling.html">
    <span class="fa fa-database"></span>
     
    Data Modeling
  </a>
</li>
<li>
  <a href="inferences.html">
    <span class="fa fa-book"></span>
     
    Data Inference
  </a>
</li>
<li>
  <a href="other.html">
    <span class="fa fa-barcode"></span>
     
    Others
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">KnowYourCpG</h1>

</div>


<p>SeSAMe implements an automated discovery tool named KnowYourCpG for discovering hidden biological and technical elements. The goal is to find a link between these elements and observed DNA methylation.</p>
<div id="knowyourcpg" class="section level1">
<h1>KnowYourCpG</h1>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>KnowYourCpG is a tool for evaluating CpG feature enrichment using Illumina probe IDs. One feature in this tool automates the hypothesis testing by asking whether a set of CpGs (represented by Illumina methylation chip probes) is enriched in certain categories or features. These categories or features can be categorical (e.g., CpGs related to tissue-specific transcription factors) or continuous (e.g., CpG Island density). Additionally, the set of CpGs to which the test will be applied can be categorical or continuous as well.</p>
<p>The set of CpGs that will be tested for enrichment is called the query set, and the set of CpGs that will be used to determine enrichment of a given factor is called the database set. A query set, for example, might be the results of an epigenome-wide association study. We have taken the time to curate our own database sets from a variety of sources that describe different categorical and continuous features such as transcription factor binding sites, CpG density, technical factors, etc.</p>
<p>Additionally, knowYourCpG has support for feature selection and feature engineering, which is currently in development.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sesame)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SummarizedExperiment)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stats4)</span></code></pre></div>
</div>
</div>
<div id="input-data" class="section level1">
<h1>Input Data</h1>
<div id="database-sets" class="section level2">
<h2>Database Sets</h2>
<p>We have organized our database sets in terms of different releases. The <code>listDatabaseSets()</code> function lists the database sets in the most recent release. The <code>release</code> parameter to this function controls which release to list. The user can see the database sets available in Release 2 using the following:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">listDatabaseSets</span>()</span></code></pre></div>
<pre><code>## Accession: KYCG.EPIC.CGI.20210713                             (n: NA)
## Accession: KYCG.EPIC.chromosome.hg19.20210630                 (n: NA)
## Accession: KYCG.EPIC.chromosome.hg38.20210630                 (n: NA)
## Accession: KYCG.EPIC.gene.20210726                            (n: NA)
## Accession: KYCG.EPIC.histoneMod.20210819                      (n: NA)
## Accession: KYCG.EPIC.mask.hg19.20181001                       (n: NA)
## Accession: KYCG.EPIC.mask.hg38.20181001                       (n: NA)
## Accession: KYCG.EPIC.probeType.20210630                       (n: NA)
## Accession: KYCG.EPIC.seqContext.20210713                      (n: NA)
## Accession: KYCG.EPIC.TFBS.20210817                            (n: NA)
## Accession: KYCG.HM450.chromosome.hg19.20210630                (n: NA)
## Accession: KYCG.HM450.gene.20210706                           (n: NA)
## Accession: KYCG.HM450.mask.hg19.20181001                      (n: NA)
## Accession: KYCG.HM450.mask.hg38.20181001                      (n: NA)
## Accession: KYCG.MM285.chromHMM.20210210                       (n: NA)
## Accession: KYCG.MM285.chromosome.mm10.20210630                (n: NA)
## Accession: KYCG.MM285.designGroup.20210210                    (n: NA)
## Accession: KYCG.MM285.gene.20210726                           (n: NA)
## Accession: KYCG.MM285.probeType.20210630                      (n: NA)
## Accession: KYCG.MM285.seqContext.20210630                     (n: NA)
## Accession: KYCG.MM285.seqContextN.20210630                    (n: NA)
## Accession: KYCG.MM285.TFBS.20210817                           (n: NA)</code></pre>
<p>We have provided the <code>getDatabaseSet()</code> function that returns a list of database sets identified by their accessions shown in <code>listDatabaseSets()</code>. When this function is ran for the first time, none of the database sets have been cached. Caching on the local machine is important on two fronts: firstly it limits the number of requests sent to our server, and secondly it limits the amount of time the user needs to wait when re-downloading database sets. For this reason, the first time the user runs <code>getDatabaseSet()</code>, it will take some time to download all of the database sets from a given release. The default directory to which the database set will be cached is in knowYourCpGPackage/databaseSets. Additionally, the user may set a global variable KYCG_DATABASESETS_LOC if they wish to specify an alternative directory. Regardless of the initial request to <code>getDatabaseSet()</code>, the entire given Release (identified by the <code>release</code> parameter) will be cached and stored in a respective sub-directory.</p>
<p>Using a subset of the available database sets, we will download them and load them in into memory using their accessions.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>databaseSetNames <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&#39;KYCG.MM285.seqContextN.20210630&#39;</span>,</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>                     <span class="st">&#39;KYCG.MM285.designGroup.20210210&#39;</span>,</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>                     <span class="st">&#39;HM450.chromosome.hg19.20210630&#39;</span>,</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>                     <span class="st">&#39;KYCG.MM285.probeType.20210630&#39;</span>)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>databaseSets <span class="ot">=</span> <span class="fu">getDatabaseSets</span>(databaseSetNames, <span class="at">verbose=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<p>In total, 39 database sets have been loaded from this command.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="fu">sprintf</span>(<span class="st">&quot;length(databaseSets) = %s database sets&quot;</span>, <span class="fu">length</span>(databaseSets)))</span></code></pre></div>
<pre><code>## length(databaseSets) = 39 database sets</code></pre>
<p>We will show the first three for simplicity.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(databaseSets[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>])</span></code></pre></div>
<pre><code>## List of 3
##  $ Enhancer: chr [1:58759] &quot;cg36603393_BC21&quot; &quot;cg36605130_BC21&quot; &quot;cg36624203_BC21&quot; &quot;cg36624767_TC21&quot; ...
##  $ Random  : chr [1:27999] &quot;cg36603791_TC21&quot; &quot;cg36608073_BC21&quot; &quot;cg36608947_BC11&quot; &quot;cg36609796_TC21&quot; ...
##  $ EPIC    : chr [1:29008] &quot;cg36603848_TC21&quot; &quot;cg36609753_BC21&quot; &quot;cg36609921_TC21&quot; &quot;cg36610885_TC21&quot; ...</code></pre>
<p>And display a summary of its contents.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(databaseSets[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>])</span></code></pre></div>
<pre><code>##          Length Class  Mode     
## Enhancer 58759  -none- character
## Random   27999  -none- character
## EPIC     29008  -none- character</code></pre>
<p>On subsequent runs of the <code>getDatabaseSet()</code> function, loading specific database sets from the same release will be much faster. These database sets will be persistent between R sessions so long as the directory to which they are downloaded is not deleted. Otherwise, the database sets will have to be downloaded again.</p>
</div>
<div id="query-sets" class="section level2">
<h2>Query Sets</h2>
<p>A query set represents probes of interest. It may either be in the form of a character vector where the values correspond to probe IDs or a named numeric vector where the names correspond to probe IDs.</p>
<p>A list of test query sets can be obtained from the following URL.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SummarizedExperiment)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>MM285.tissueSignature <span class="ot">=</span> <span class="fu">sesameDataGet</span>(<span class="st">&#39;MM285.tissueSignature&#39;</span>)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>df <span class="ot">=</span> <span class="fu">rowData</span>(MM285.tissueSignature)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>querySet <span class="ot">=</span> df<span class="sc">$</span>Probe_ID[df<span class="sc">$</span>branch <span class="sc">==</span> <span class="st">&quot;E-Brain&quot;</span>]</span></code></pre></div>
<p>This query set represents hypomethylated probes in Mouse brain tissue from the MM285 platform. This specific query set has 71 probes.</p>
<pre><code>## length(querySet) = 71 probes</code></pre>
<p>Using the obtained database sets and query set, the user may analyze their overlap and enrichment.</p>
</div>
</div>
<div id="functionality" class="section level1">
<h1>Functionality</h1>
<p>There are many functions in this package that make investigating the biological significance of a set of probes easier.</p>
<div id="obtaining-optional-annotation-from-each-databse-set" class="section level2">
<h2>Obtaining Optional Annotation from each Databse Set</h2>
<p>Using the <code>getDatabaseSetOverlap()</code> function, the user can investigate the database sets that overlap with the given query set, the degree of overlap, and any provided annotation.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>annotation <span class="ot">=</span> <span class="fu">getDatabaseSetOverlap</span>(querySet, databaseSets)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(annotation)</span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["meta"],"name":[1],"type":["lgl"],"align":["right"]},{"label":["nQ"],"name":[2],"type":["int"],"align":["right"]},{"label":["nD"],"name":[3],"type":["int"],"align":["right"]},{"label":["overlap"],"name":[4],"type":["int"],"align":["right"]}],"data":[{"1":"FALSE","2":"71","3":"58759","4":"34","_rn_":"Enhancer"},{"1":"FALSE","2":"71","3":"27999","4":"5","_rn_":"Random"},{"1":"FALSE","2":"71","3":"29008","4":"11","_rn_":"EPIC"},{"1":"FALSE","2":"71","3":"1486","4":"0","_rn_":"SNP"},{"1":"FALSE","2":"71","3":"8311","4":"0","_rn_":"Adenoma"},{"1":"FALSE","2":"71","3":"24638","4":"3","_rn_":"GeneBody"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>The <code>meta</code> column indicates whether the respective database set contains annotation. The <code>nQ</code> and <code>nD</code> columns indicate the length of the query set and database set respectively, and the <code>overlap</code> column indicates the amount of overlap. In some cases, this annotation data.frame will be sparse as not all database sets have the same annotation columns. None of these overlapping database sets have annotation, so no additional columns were returned.</p>
</div>
<div id="investigating-the-enrichment-of-many-database-sets-in-a-single-query-set" class="section level2">
<h2>Investigating the Enrichment of Many Database Sets in a Single Query Set</h2>
<p>The <code>testEnrichmentAll()</code> function is the main work horse to KnowYourCpG. It tests the enrichment of the given database sets in a single query set. There are four testing scenarios depending on the type format of the query set and database sets. They are shown with the respective testing scenario in the table below.</p>
<div class="figure">
<img src="../man/figures/20210627_kycpg_tests.png" style="width:75.0%" alt="" />
<p class="caption">Four testing scenarios of KnowYourCpG</p>
</div>
<p>The <code>testEnrichmentAll()</code> will automate statistical tests and report metrics about each of the the loaded database sets. Another set that is needed for the test is called the universe set. This is the set of all probes for a given platform. It can either be passed in as an argument called <code>universeSet</code> or the platform name can be passed with argument <code>platform</code>. If neither of these are supplied, the universe set will be implied from the probes. In all subsequent runs of <code>testEnrichmentAll()</code> in this vignette, the platform will be specified.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>resultsAll <span class="ot">=</span> <span class="fu">testEnrichmentAll</span>(<span class="at">querySet=</span>querySet, </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>                               <span class="at">databaseSets=</span>databaseSets, </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>                               <span class="at">verbose=</span><span class="cn">FALSE</span>)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">head</span>(resultsAll))</span></code></pre></div>
<pre><code>##                   estimate      p.value   test nQ     nD overlap p.adjust.fdr
## Enhancer         2.4238684 9.215605e-08 fisher 71  58759      34 3.594086e-06
## cg               1.0440670 7.690410e-02 fisher 71 284862      71 1.000000e+00
## EPIC             1.5884715 1.082081e-01 fisher 71  29008      11 1.000000e+00
## CGI              0.2444814 1.930669e-01 fisher 71  17134       1 1.000000e+00
## PlacentaIntermed 4.2700751 2.091150e-01 fisher 71    981       1 1.000000e+00
## CpGDesity50      0.3822945 2.137862e-01  fgsea 71 297415      71 1.000000e+00
##                  estimate.rank p.value.rank overlap.rank max.rank mean.rank
## Enhancer                     2            1           36       36     18.75
## cg                           5            2           37       37     20.25
## EPIC                         3            3           34       34     18.50
## CGI                         12            4           28       28     18.00
## PlacentaIntermed             1            5           29       29     16.00
## CpGDesity50                 11            6           38       38     23.25
##                   meta
## Enhancer         FALSE
## cg               FALSE
## EPIC             FALSE
## CGI              FALSE
## PlacentaIntermed FALSE
## CpGDesity50      FALSE</code></pre>
<p>The output of each test contains at least four variables: the estimate, p-value, type of test, and whether meta data is included in the tested database set. The name of the database set is also recorded as well. By default, the p-value column is sorted.</p>
<p>It should be noted that the estimate (or test statistic) is test dependent and comparison between p-values should be limited to within the same type of test. For instance, the test statistics for Fisher’s exact test and FGSEA are log fold change and the test statistic for Spearman’s test is simply the rank order correlation coefficient. For simplicity, we report all of the test types in one data frame.</p>
<p>The <code>nQ</code> and <code>nD</code> columns identify the length of the query set and the database set, respectively. Often, it’s important to examine the extent of overlap between the two sets, so that metric is reported as well in the <code>overlap</code> column.</p>
<p>Using these results, we can plot a volcano plot and lollipop plot.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotVolcano</span>(<span class="at">data=</span>resultsAll, <span class="at">title=</span><span class="st">&quot;Database Set Enrichment&quot;</span>, </span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">subtitle=</span><span class="st">&quot;MM285 Mouse Platform&quot;</span>)</span></code></pre></div>
<p><img src="kyCG_files/figure-html/plot-volcano-1.png" width="672" /></p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotLollipop</span>(<span class="at">data=</span>resultsAll, <span class="at">title=</span><span class="st">&quot;Top Database Set Enrichment&quot;</span>, </span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">subtitle=</span><span class="st">&quot;MM285 Mouse Platform&quot;</span>)</span></code></pre></div>
<p><img src="kyCG_files/figure-html/plot-lollipop-1.png" width="672" /></p>
</div>
<div id="example-testing-scenarios" class="section level2">
<h2>Example Testing Scenarios</h2>
<p>The querySet may be a named continuous vector. In that case, either a gene enrichment score will be calculated (if the databaseSet is discrete) or a Spearman correlation will be calculated (if the databaseSet is continuous as well). The three other cases are shown below using biologically relevant examples.</p>
<p>To display this functionality, let’s load two numeric database sets individually. One is a database set for CpG density and the other is a database set corresponding to the distance of the nearest transcriptional start site (TSS) to each probe.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>KYCG.MM285.seqContextN<span class="fl">.20210630</span> <span class="ot">=</span> <span class="fu">getDatabaseSets</span>(<span class="st">&#39;KYCG.MM285.seqContextN.20210630&#39;</span>, <span class="at">verbose=</span><span class="cn">FALSE</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>CpGDesity50 <span class="ot">=</span> KYCG.MM285.seqContextN<span class="fl">.20210630</span>[<span class="st">&#39;CpGDesity50&#39;</span>]</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>distToTSS <span class="ot">=</span> KYCG.MM285.seqContextN<span class="fl">.20210630</span>[<span class="st">&#39;distToTSS&#39;</span>]</span></code></pre></div>
<div id="example-1" class="section level3">
<h3>Example 1</h3>
<p>Our first test will look at whether the query set is significantly enriched in either of the database sets individually.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>resultsCpGDensity <span class="ot">=</span> <span class="fu">testEnrichmentAll</span>(<span class="at">querySet=</span>querySet,</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">databaseSets=</span>CpGDesity50,</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">platform=</span><span class="st">&quot;MM285&quot;</span>)</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(resultsCpGDensity)</span></code></pre></div>
<pre><code>##              estimate   p.value  test nQ     nD overlap p.adjust.fdr
## CpGDesity50 0.3822945 0.2417582 fgsea 71 297415      71    0.2417582
##             estimate.rank p.value.rank overlap.rank max.rank mean.rank  meta
## CpGDesity50             1            1            1        1         1 FALSE</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>resultsTSS <span class="ot">=</span> <span class="fu">testEnrichmentAll</span>(<span class="at">querySet=</span>querySet, </span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>                               <span class="at">databaseSets=</span>distToTSS, </span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>                               <span class="at">platform=</span><span class="st">&quot;MM285&quot;</span>)</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(resultsTSS)</span></code></pre></div>
<pre><code>##            estimate   p.value  test nQ     nD overlap p.adjust.fdr
## distToTSS 0.5857876 0.9210789 fgsea 71 303421      71    0.9210789
##           estimate.rank p.value.rank overlap.rank max.rank mean.rank  meta
## distToTSS             1            1            1        1         1 FALSE</code></pre>
<p>Both of the tests result in a p-value greater than 0.05, so there is little evidence to suggest that either of the query set is enriched in either of these database sets.</p>
</div>
<div id="example-2" class="section level3">
<h3>Example 2</h3>
<p>A third test might be whether there is a significant correlation between these two database sets.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>resultsCpGdensityTSS <span class="ot">=</span> <span class="fu">testEnrichmentAll</span>(<span class="at">querySet=</span>CpGDesity50,</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">databaseSets=</span>distToTSS,</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">platform=</span><span class="st">&quot;MM285&quot;</span>)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(resultsCpGdensityTSS)</span></code></pre></div>
<pre><code>##           estimate p.value  test nQ     nD overlap p.adjust.fdr estimate.rank
## distToTSS        0       1 fgsea  1 303421       0            1             1
##           p.value.rank overlap.rank max.rank mean.rank  meta
## distToTSS            1            1        1         1 FALSE</code></pre>
<p>This test is statistically significant with a p-value less than 0.05 and a moderately weak negative correlation. It is known in biology that the distance to TSS are negatively correlated with CpG islands, which is shown to be the case here from the data as well.</p>
<p>In the above four examples, the database sets were specified explicitly. This is not necessarily needed as the function can load a default set of database sets based a release specified the <code>release</code> parameter. The default is <code>release=2</code>.</p>
</div>
<div id="example-3" class="section level3">
<h3>Example 3</h3>
<p>We may want to look specifically at query our query set is significantly enriched in the set of all Transcription Factor Binding Sites (TFBS). Let’s first load in the TFBS database set.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>databaseSets <span class="ot">=</span> <span class="fu">getDatabaseSets</span>(<span class="st">&quot;KYCG.MM285.TFBS.20210817&quot;</span>, <span class="at">verbose=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<p>Next we can actually perform the test.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>resultsTFBS <span class="ot">=</span> <span class="fu">testEnrichmentAll</span>(<span class="at">querySet=</span>querySet, <span class="at">databaseSets=</span>databaseSets,</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>                                <span class="at">platform=</span><span class="st">&quot;MM285&quot;</span>, <span class="at">verbose=</span><span class="cn">FALSE</span>)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(resultsTFBS)</span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["estimate"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["p.value"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["test"],"name":[3],"type":["chr"],"align":["left"]},{"label":["nQ"],"name":[4],"type":["int"],"align":["right"]},{"label":["nD"],"name":[5],"type":["int"],"align":["right"]},{"label":["overlap"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["p.adjust.fdr"],"name":[7],"type":["dbl"],"align":["right"]},{"label":["estimate.rank"],"name":[8],"type":["int"],"align":["right"]},{"label":["p.value.rank"],"name":[9],"type":["int"],"align":["right"]},{"label":["overlap.rank"],"name":[10],"type":["int"],"align":["right"]},{"label":["max.rank"],"name":[11],"type":["int"],"align":["right"]},{"label":["mean.rank"],"name":[12],"type":["dbl"],"align":["right"]},{"label":["meta"],"name":[13],"type":["named list"],"align":["right"]},{"label":["FileAccession"],"name":[14],"type":["named list"],"align":["right"]},{"label":["Target"],"name":[15],"type":["named list"],"align":["right"]},{"label":["TargetGroup"],"name":[16],"type":["named list"],"align":["right"]},{"label":["Format"],"name":[17],"type":["named list"],"align":["right"]},{"label":["Tissue"],"name":[18],"type":["named list"],"align":["right"]},{"label":["CellLine"],"name":[19],"type":["named list"],"align":["right"]},{"label":["dbID"],"name":[20],"type":["named list"],"align":["right"]}],"data":[{"1":"19.83112","2":"1.218297e-13","3":"fisher","4":"71","5":"2746","6":"13","7":"1.591882e-10","8":"12","9":"1","10":"1341","11":"1341","12":"673.75","13":"<lgl [1]>","14":"<chr [1]>","15":"<chr [1]>","16":"<chr [1]>","17":"<chr [1]>","18":"<chr [1]>","19":"<chr [1]>","20":"<chr [1]>","_rn_":"GSM1629386"},{"1":"10.82744","2":"2.360092e-13","3":"fisher","4":"71","5":"6577","6":"17","7":"1.591882e-10","8":"19","9":"2","10":"1347","11":"1347","12":"678.75","13":"<lgl [1]>","14":"<chr [1]>","15":"<chr [1]>","16":"<chr [1]>","17":"<chr [1]>","18":"<chr [1]>","19":"<chr [1]>","20":"<chr [1]>","_rn_":"GSM1629388"},{"1":"19.79652","2":"1.013709e-10","3":"fisher","4":"71","5":"2116","6":"10","7":"4.558312e-08","8":"13","9":"3","10":"1339","11":"1339","12":"673.50","13":"<lgl [1]>","14":"<chr [1]>","15":"<chr [1]>","16":"<chr [1]>","17":"<chr [1]>","18":"<chr [1]>","19":"<chr [1]>","20":"<chr [1]>","_rn_":"GSM1553879"},{"1":"9.66945","2":"1.714858e-10","3":"fisher","4":"71","5":"6065","6":"14","7":"5.783359e-08","8":"22","9":"4","10":"1342","11":"1342","12":"677.50","13":"<lgl [1]>","14":"<chr [1]>","15":"<chr [1]>","16":"<chr [1]>","17":"<chr [1]>","18":"<chr [1]>","19":"<chr [1]>","20":"<chr [1]>","_rn_":"GSM1944758"},{"1":"10.15786","2":"4.585502e-10","3":"fisher","4":"71","5":"5361","6":"13","7":"1.237168e-07","8":"20","9":"5","10":"1340","11":"1340","12":"676.25","13":"<lgl [1]>","14":"<chr [1]>","15":"<chr [1]>","16":"<chr [1]>","17":"<chr [1]>","18":"<chr [1]>","19":"<chr [1]>","20":"<chr [1]>","_rn_":"GSM782847"},{"1":"20.63519","2":"6.095443e-09","3":"fisher","4":"71","5":"1624","6":"8","7":"1.370459e-06","8":"9","9":"6","10":"1331","11":"1331","12":"669.25","13":"<lgl [1]>","14":"<chr [1]>","15":"<chr [1]>","16":"<chr [1]>","17":"<chr [1]>","18":"<chr [1]>","19":"<chr [1]>","20":"<chr [1]>","_rn_":"GSM1824088"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Using these sample results, we can plot a volcano plot and lollipop plot.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotVolcano</span>(<span class="at">data=</span>resultsTFBS, <span class="at">title=</span><span class="st">&quot;Transcription Factor Binding Site Enrichment&quot;</span>,</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">subtitle=</span><span class="st">&#39;MM285 Mouse Platform&#39;</span>)</span></code></pre></div>
<pre><code>## Warning: ggrepel: 44 unlabeled data points (too many overlaps). Consider
## increasing max.overlaps</code></pre>
<p><img src="kyCG_files/figure-html/plot-volcano-tfbs-1.png" width="672" /></p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotLollipop</span>(<span class="at">data=</span>resultsTFBS, <span class="at">title=</span><span class="st">&quot;Transcription Factor Binding Site Enrichment&quot;</span>,</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">subtitle=</span><span class="st">&#39;MM285 Mouse Platform&#39;</span>)</span></code></pre></div>
<p><img src="kyCG_files/figure-html/plot-lollipop-tfbs-1.png" width="672" /></p>
</div>
</div>
<div id="gene-enrichment-ananlysis" class="section level2">
<h2>Gene Enrichment Ananlysis</h2>
<p>Automating the enrichment test process only works when the number of database sets is small. This is important when targeting all genes as there are tens of thousands of genes on each platform. By testing only those genes that overlap with the query set, we can greatly reduce the number of tests. For this reason, the gene enrichment analysis is a special case of these enrichment tests. We can perform this analysis using the <code>testEnrichmentGene()</code> function.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>resultsGene <span class="ot">=</span> <span class="fu">testEnrichmentGene</span>(querySet, <span class="at">platform=</span><span class="st">&quot;MM285&quot;</span>, <span class="at">verbose=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>## [1] &quot;Retrieving 0 databaseSets...&quot;</code></pre>
<pre><code>## ExperimentHub not responding. Using backup.</code></pre>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(resultsGene)</span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["estimate"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["p.value"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["test"],"name":[3],"type":["chr"],"align":["left"]},{"label":["nQ"],"name":[4],"type":["int"],"align":["right"]},{"label":["nD"],"name":[5],"type":["int"],"align":["right"]},{"label":["overlap"],"name":[6],"type":["int"],"align":["right"]},{"label":["p.adjust.fdr"],"name":[7],"type":["dbl"],"align":["right"]},{"label":["estimate.rank"],"name":[8],"type":["int"],"align":["right"]},{"label":["p.value.rank"],"name":[9],"type":["int"],"align":["right"]},{"label":["overlap.rank"],"name":[10],"type":["int"],"align":["right"]},{"label":["max.rank"],"name":[11],"type":["int"],"align":["right"]},{"label":["mean.rank"],"name":[12],"type":["dbl"],"align":["right"]},{"label":["meta"],"name":[13],"type":["lgl"],"align":["right"]}],"data":[{"1":"571.2196","2":"2.000907e-08","3":"fisher","4":"71","5":"22","6":"3","7":"0.0005601186","8":"15","9":"1","10":"59","11":"59","12":"33.50","13":"FALSE","_rn_":"D030068K23Rik"},{"1":"523.6180","2":"2.628862e-08","3":"fisher","4":"71","5":"24","6":"3","7":"0.0005601186","8":"17","9":"2","10":"58","11":"58","12":"33.75","13":"FALSE","_rn_":"Fam110b"},{"1":"418.8944","2":"5.267889e-08","3":"fisher","4":"71","5":"30","6":"3","7":"0.0007482685","8":"23","9":"3","10":"60","11":"60","12":"36.50","13":"FALSE","_rn_":"Map4"},{"1":"761.6261","2":"3.085960e-06","3":"fisher","4":"71","5":"11","6":"2","7":"0.0328755006","8":"11","9":"4","10":"57","11":"57","12":"32.25","13":"FALSE","_rn_":"Tnks"},{"1":"465.4382","2":"8.575291e-06","3":"fisher","4":"71","5":"18","6":"2","7":"0.0680576956","8":"21","9":"5","10":"56","11":"56","12":"34.50","13":"FALSE","_rn_":"Rufy3"},{"1":"440.9414","2":"9.582667e-06","3":"fisher","4":"71","5":"19","6":"2","7":"0.0680576956","8":"22","9":"6","10":"54","11":"54","12":"34.00","13":"FALSE","_rn_":"Myt1"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Using these sample results, we can plot a volcano plot and lollipop plot.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotVolcano</span>(<span class="at">data=</span>resultsGene, <span class="at">title=</span><span class="st">&quot;Gene Enrichment&quot;</span>,</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">subtitle=</span><span class="st">&quot;MM285 Mouse Platform&quot;</span>, <span class="at">n.fdr=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="kyCG_files/figure-html/plot-volcano-gene-1.png" width="672" /></p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotLollipop</span>(<span class="at">data=</span>resultsGene, <span class="at">title=</span><span class="st">&quot;Top Gene Enrichment&quot;</span>, </span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>             <span class="at">subtitle=</span><span class="st">&quot;MM285 Mouse Platform&quot;</span>, <span class="at">n=</span><span class="dv">10</span>)</span></code></pre></div>
<p><img src="kyCG_files/figure-html/plot-lollipop-gene-1.png" width="672" /></p>
<p>For example, this given query set is tissue specific hypomethylation of mouse brain. Rufy3 is shown to be significantly enriched in this set and it is known to be enriched in neurons (<a href="https://www.ncbi.nlm.nih.gov/gene/22902" class="uri">https://www.ncbi.nlm.nih.gov/gene/22902</a>).</p>
</div>
</div>
<div id="feature-engineering" class="section level1">
<h1>Feature Engineering</h1>
<p>In addition to hypothesis testing, knowYourCpG also uses the curated database sets for feature engineering. We have a pre-curated summarized experiment containing a samplesheet and beta value matrix corresponding to about 467 MM285 samples with 20k probes. The samplesheet includes UIDs pertaining to the sample and several categorical/numerical features. To use this data for a linear model, we will extract the most relevant prevalent features.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>se <span class="ot">=</span> <span class="fu">sesameDataGet</span>(<span class="st">&#39;MM285.20Kx467.SE&#39;</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>samplesheet <span class="ot">=</span> <span class="fu">colData</span>(se)[, <span class="fu">c</span>(<span class="st">&quot;Mouse_Age_Months&quot;</span>,</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>         <span class="st">&quot;Mouse_Age_Days&quot;</span>, <span class="st">&quot;Sex&quot;</span>, <span class="st">&quot;Strain_Corrected&quot;</span>,</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>         <span class="st">&quot;Tissue_Corrected&quot;</span>, <span class="st">&#39;Genotype&#39;</span>)]</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>betas <span class="ot">=</span> <span class="fu">assay</span>(se)</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">head</span>(samplesheet))</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 6 columns
##   Mouse_Age_Months Mouse_Age_Days         Sex Strain_Corrected Tissue_Corrected
##        &lt;character&gt;    &lt;character&gt; &lt;character&gt;      &lt;character&gt;      &lt;character&gt;
## 1                1             29      Female           129/Sv            Colon
## 2                1             31        Male           129/Sv            Colon
## 3             17.5            531      Female           129/Sv            Colon
## 4               16            497        Male           129/Sv            Colon
## 5                1             29      Female           129/Sv            Colon
## 6               18            542      Female           129/Sv            Colon
##      Genotype
##   &lt;character&gt;
## 1          WT
## 2          WT
## 3          WT
## 4          WT
## 5          WT
## 6          WT</code></pre>
<p>We have found that it is incredibly computationally expensive to perform a linear model/generalized linear model on a feature set of individual CpGs. Additionally, interpreting the mechanism the significantly contributing CpGs is non-trivial due to their complex interactions. We hope to leverage these pre-curated database sets by using their beta value summary statistics as features instead.</p>
<p>We will calculate the summary statistics for the betas matrix using a list of database sets.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>databaseSets <span class="ot">=</span> <span class="fu">getDatabaseSets</span>(databaseSetNames, <span class="at">verbose=</span><span class="cn">FALSE</span>)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>statistics <span class="ot">=</span> <span class="fu">calcDatabaseSetStatisticsAll</span>(betas, <span class="at">databaseSets=</span>databaseSets)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(statistics[, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>])</span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SNP-mean"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["SNP-median"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["SNP-var"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["SNP-sd"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["SNP-skew"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"0.6087093","2":"0.9303616","3":"0.1879842","4":"0.4335715","5":"0.1879842","_rn_":"1"},{"1":"0.6134781","2":"0.9046590","3":"0.1758904","4":"0.4193929","5":"0.1758904","_rn_":"2"},{"1":"0.6078863","2":"0.9236646","3":"0.1836084","4":"0.4284955","5":"0.1836084","_rn_":"3"},{"1":"0.6108740","2":"0.9235892","3":"0.1833909","4":"0.4282416","5":"0.1833909","_rn_":"4"},{"1":"0.6089673","2":"0.9256238","3":"0.1865235","4":"0.4318837","5":"0.1865235","_rn_":"5"},{"1":"0.6097658","2":"0.9281700","3":"0.1860721","4":"0.4313608","5":"0.1860721","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>We will look at the mean of each database set.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>statistics <span class="ot">=</span> statistics[, <span class="fu">grepl</span>(<span class="st">&quot;mean&quot;</span>, <span class="fu">colnames</span>(statistics))]</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(statistics[, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>])</span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["SNP-mean"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["Adenoma-mean"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["CGI-mean"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["CTCF-mean"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["OocyteMeth-mean"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"0.6087093","2":"0.5967435","3":"0.1858567","4":"0.2914830","5":"0.03812385","_rn_":"1"},{"1":"0.6134781","2":"0.6391305","3":"0.1961624","4":"0.2920514","5":"0.07895942","_rn_":"2"},{"1":"0.6078863","2":"0.5646461","3":"0.2261244","4":"0.3095448","5":"0.05740676","_rn_":"3"},{"1":"0.6108740","2":"0.6026663","3":"0.2123136","4":"0.3048324","5":"0.07298051","_rn_":"4"},{"1":"0.6089673","2":"0.5955713","3":"0.1848392","4":"0.2871704","5":"0.03931924","_rn_":"5"},{"1":"0.6097658","2":"0.6331576","3":"0.2174764","4":"0.3217050","5":"0.08692195","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Just from the few database set means above, we can see that CpG islands are consistently hypomethylated, which is consistent with known biology.</p>
<p>Using the samplesheet and beta values, we can create a singular data for linear models. Each categorical variable should be caste as a factor with a reference level. Each numerical variable should be expressed appropriately as well.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">=</span> <span class="fu">cbind</span>(<span class="fu">data.frame</span>(samplesheet), statistics)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>Sex <span class="ot">=</span> <span class="fu">relevel</span>(<span class="fu">factor</span>(data<span class="sc">$</span>Sex), <span class="st">&#39;Female&#39;</span>)</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>Strain_Corrected <span class="ot">=</span> <span class="fu">relevel</span>(<span class="fu">factor</span>(data<span class="sc">$</span>Strain_Corrected), <span class="st">&#39;129/Sv&#39;</span>)</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>Tissue_Corrected <span class="ot">=</span> <span class="fu">relevel</span>(<span class="fu">factor</span>(data<span class="sc">$</span>Tissue_Corrected),  <span class="st">&#39;Colon&#39;</span>)</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>Genotype <span class="ot">=</span> <span class="fu">relevel</span>(<span class="fu">factor</span>(data<span class="sc">$</span>Genotype), <span class="st">&#39;WT&#39;</span>)</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>Mouse_Age_Days <span class="ot">=</span> <span class="fu">as.numeric</span>(data<span class="sc">$</span>Mouse_Age_Days)</span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>data<span class="sc">$</span>Mouse_Age_Months <span class="ot">=</span> <span class="fu">as.numeric</span>(data<span class="sc">$</span>Mouse_Age_Months)</span></code></pre></div>
<p>A linear model can finally be constructed using this data. Since there are only 20 samples, we will use use four of the database summary statistics.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">=</span> <span class="fu">lm</span>(Mouse_Age_Days <span class="sc">~</span> Sex <span class="sc">+</span> Strain_Corrected <span class="sc">+</span> Tissue_Corrected <span class="sc">+</span> </span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>               Genotype <span class="sc">+</span> <span class="st">`</span><span class="at">CGI-mean</span><span class="st">`</span> <span class="sc">+</span> <span class="st">`</span><span class="at">CTCF-mean</span><span class="st">`</span> <span class="sc">+</span> <span class="st">`</span><span class="at">Clock-mean</span><span class="st">`</span> <span class="sc">+</span> </span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>               <span class="st">`</span><span class="at">SNP-mean</span><span class="st">`</span> <span class="sc">+</span> <span class="st">`</span><span class="at">SpermMeth-mean</span><span class="st">`</span> <span class="sc">+</span> <span class="st">`</span><span class="at">VMR-mean</span><span class="st">`</span>, <span class="at">data=</span>data)</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(model)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = Mouse_Age_Days ~ Sex + Strain_Corrected + Tissue_Corrected + 
##     Genotype + `CGI-mean` + `CTCF-mean` + `Clock-mean` + `SNP-mean` + 
##     `SpermMeth-mean` + `VMR-mean`, data = data)
## 
## Residuals:
##     Min      1Q  Median      3Q     Max 
## -560.89  -53.44   -5.61   38.85  506.67 
## 
## Coefficients: (1 not defined because of singularities)
##                                                                                        Estimate
## (Intercept)                                                                            -2380.62
## SexMale                                                                                   60.06
## Strain_Corrected129-C57 Mixture                                                           41.64
## Strain_Corrected129S1/SvlmJ                                                             -131.32
## Strain_CorrectedA/J                                                                     -163.98
## Strain_CorrectedAKR/J                                                                   -109.23
## Strain_CorrectedAthymic_Nude                                                            -248.68
## Strain_CorrectedBALB/cJ                                                                 -124.78
## Strain_CorrectedBTBR_T+ ITpr3tf/J                                                       -148.92
## Strain_CorrectedC3H/HeJ                                                                 -131.69
## Strain_CorrectedC57BL/10J                                                               -186.93
## Strain_CorrectedC57BL/6J                                                                  81.84
## Strain_CorrectedC57BL/6NJ                                                               -194.02
## Strain_CorrectedC57L/J                                                                  -103.21
## Strain_CorrectedC58/J                                                                   -218.37
## Strain_CorrectedCAST/EiJ                                                                -198.39
## Strain_CorrectedCBA/J                                                                   -151.89
## Strain_CorrectedDBA/1J                                                                  -113.94
## Strain_CorrectedDBA/2J                                                                  -136.09
## Strain_CorrectedFVB/NJ                                                                  -178.02
## Strain_CorrectedKK/HiJ                                                                  -203.66
## Strain_CorrectedLP/J                                                                    -153.80
## Strain_CorrectedMixed                                                                     74.85
## Strain_CorrectedMOLF/EiJ                                                                -291.90
## Strain_CorrectedNOD/ShiLtJ                                                              -176.53
## Strain_CorrectedNSG                                                                     -219.59
## Strain_CorrectedNZB/BINJ                                                                -156.81
## Strain_CorrectedNZO/HlLtJ                                                               -169.46
## Strain_CorrectedNZW/LacJ                                                                -174.28
## Strain_CorrectedPWK/PhJ                                                                 -292.74
## Strain_CorrectedWSB/EiJ                                                                  -90.21
## Tissue_CorrectedCecum                                                                    -66.86
## Tissue_CorrectedEsophagus                                                                 46.50
## Tissue_CorrectedFat                                                                      117.13
## Tissue_CorrectedFetal Brain                                                               74.62
## Tissue_CorrectedFetal Intestine                                                           76.42
## Tissue_CorrectedFetal Limb                                                                54.05
## Tissue_CorrectedFrontal Lobe Brain                                                       109.26
## Tissue_CorrectedHeart                                                                    109.46
## Tissue_CorrectedHind Brain                                                               198.13
## Tissue_CorrectedKidney                                                                   172.79
## Tissue_CorrectedLiver                                                                    -26.65
## Tissue_CorrectedLung                                                                     126.91
## Tissue_CorrectedMuscle                                                                   117.13
## Tissue_CorrectedOvary                                                                    131.78
## Tissue_CorrectedPlacenta                                                                -446.97
## Tissue_CorrectedSmall Intestine                                                          -42.09
## Tissue_CorrectedSpleen                                                                  -141.38
## Tissue_CorrectedStomach                                                                  172.22
## Tissue_CorrectedTail                                                                      80.72
## Tissue_CorrectedTestis                                                                  -775.34
## GenotypeAPC:min/+  ApclacIGY:ApclacIGY/+  Dnmt1:LO/LO VilLacIGY:+/+  Villin-mKate2:+/+       NA
## `CGI-mean`                                                                              8261.61
## `CTCF-mean`                                                                            -3023.30
## `Clock-mean`                                                                            3753.33
## `SNP-mean`                                                                              1771.94
## `SpermMeth-mean`                                                                        3650.00
## `VMR-mean`                                                                             -2639.44
##                                                                                        Std. Error
## (Intercept)                                                                               1028.40
## SexMale                                                                                     15.34
## Strain_Corrected129-C57 Mixture                                                             53.27
## Strain_Corrected129S1/SvlmJ                                                                 51.40
## Strain_CorrectedA/J                                                                         57.68
## Strain_CorrectedAKR/J                                                                       59.60
## Strain_CorrectedAthymic_Nude                                                               138.28
## Strain_CorrectedBALB/cJ                                                                     50.75
## Strain_CorrectedBTBR_T+ ITpr3tf/J                                                           51.72
## Strain_CorrectedC3H/HeJ                                                                     53.75
## Strain_CorrectedC57BL/10J                                                                   55.20
## Strain_CorrectedC57BL/6J                                                                    24.93
## Strain_CorrectedC57BL/6NJ                                                                   52.54
## Strain_CorrectedC57L/J                                                                      69.01
## Strain_CorrectedC58/J                                                                       56.65
## Strain_CorrectedCAST/EiJ                                                                    83.09
## Strain_CorrectedCBA/J                                                                       49.41
## Strain_CorrectedDBA/1J                                                                      49.60
## Strain_CorrectedDBA/2J                                                                      49.29
## Strain_CorrectedFVB/NJ                                                                      54.77
## Strain_CorrectedKK/HiJ                                                                      59.68
## Strain_CorrectedLP/J                                                                        62.12
## Strain_CorrectedMixed                                                                       45.35
## Strain_CorrectedMOLF/EiJ                                                                    71.70
## Strain_CorrectedNOD/ShiLtJ                                                                  51.83
## Strain_CorrectedNSG                                                                        138.60
## Strain_CorrectedNZB/BINJ                                                                    55.39
## Strain_CorrectedNZO/HlLtJ                                                                   51.93
## Strain_CorrectedNZW/LacJ                                                                    52.82
## Strain_CorrectedPWK/PhJ                                                                     73.27
## Strain_CorrectedWSB/EiJ                                                                     53.94
## Tissue_CorrectedCecum                                                                       57.56
## Tissue_CorrectedEsophagus                                                                   55.59
## Tissue_CorrectedFat                                                                         61.99
## Tissue_CorrectedFetal Brain                                                                 63.46
## Tissue_CorrectedFetal Intestine                                                             61.11
## Tissue_CorrectedFetal Limb                                                                  59.41
## Tissue_CorrectedFrontal Lobe Brain                                                          44.89
## Tissue_CorrectedHeart                                                                       52.03
## Tissue_CorrectedHind Brain                                                                  61.61
## Tissue_CorrectedKidney                                                                      49.95
## Tissue_CorrectedLiver                                                                       38.45
## Tissue_CorrectedLung                                                                        46.89
## Tissue_CorrectedMuscle                                                                      61.86
## Tissue_CorrectedOvary                                                                       84.15
## Tissue_CorrectedPlacenta                                                                    59.33
## Tissue_CorrectedSmall Intestine                                                             37.31
## Tissue_CorrectedSpleen                                                                      49.74
## Tissue_CorrectedStomach                                                                     43.12
## Tissue_CorrectedTail                                                                        46.33
## Tissue_CorrectedTestis                                                                     140.85
## GenotypeAPC:min/+  ApclacIGY:ApclacIGY/+  Dnmt1:LO/LO VilLacIGY:+/+  Villin-mKate2:+/+         NA
## `CGI-mean`                                                                                 851.87
## `CTCF-mean`                                                                               1030.30
## `Clock-mean`                                                                               747.98
## `SNP-mean`                                                                                1685.20
## `SpermMeth-mean`                                                                           515.44
## `VMR-mean`                                                                                 818.24
##                                                                                        t value
## (Intercept)                                                                             -2.315
## SexMale                                                                                  3.915
## Strain_Corrected129-C57 Mixture                                                          0.782
## Strain_Corrected129S1/SvlmJ                                                             -2.555
## Strain_CorrectedA/J                                                                     -2.843
## Strain_CorrectedAKR/J                                                                   -1.833
## Strain_CorrectedAthymic_Nude                                                            -1.798
## Strain_CorrectedBALB/cJ                                                                 -2.459
## Strain_CorrectedBTBR_T+ ITpr3tf/J                                                       -2.879
## Strain_CorrectedC3H/HeJ                                                                 -2.450
## Strain_CorrectedC57BL/10J                                                               -3.387
## Strain_CorrectedC57BL/6J                                                                 3.283
## Strain_CorrectedC57BL/6NJ                                                               -3.693
## Strain_CorrectedC57L/J                                                                  -1.496
## Strain_CorrectedC58/J                                                                   -3.855
## Strain_CorrectedCAST/EiJ                                                                -2.388
## Strain_CorrectedCBA/J                                                                   -3.074
## Strain_CorrectedDBA/1J                                                                  -2.297
## Strain_CorrectedDBA/2J                                                                  -2.761
## Strain_CorrectedFVB/NJ                                                                  -3.250
## Strain_CorrectedKK/HiJ                                                                  -3.413
## Strain_CorrectedLP/J                                                                    -2.476
## Strain_CorrectedMixed                                                                    1.651
## Strain_CorrectedMOLF/EiJ                                                                -4.071
## Strain_CorrectedNOD/ShiLtJ                                                              -3.406
## Strain_CorrectedNSG                                                                     -1.584
## Strain_CorrectedNZB/BINJ                                                                -2.831
## Strain_CorrectedNZO/HlLtJ                                                               -3.263
## Strain_CorrectedNZW/LacJ                                                                -3.299
## Strain_CorrectedPWK/PhJ                                                                 -3.996
## Strain_CorrectedWSB/EiJ                                                                 -1.672
## Tissue_CorrectedCecum                                                                   -1.162
## Tissue_CorrectedEsophagus                                                                0.836
## Tissue_CorrectedFat                                                                      1.890
## Tissue_CorrectedFetal Brain                                                              1.176
## Tissue_CorrectedFetal Intestine                                                          1.251
## Tissue_CorrectedFetal Limb                                                               0.910
## Tissue_CorrectedFrontal Lobe Brain                                                       2.434
## Tissue_CorrectedHeart                                                                    2.104
## Tissue_CorrectedHind Brain                                                               3.216
## Tissue_CorrectedKidney                                                                   3.459
## Tissue_CorrectedLiver                                                                   -0.693
## Tissue_CorrectedLung                                                                     2.706
## Tissue_CorrectedMuscle                                                                   1.894
## Tissue_CorrectedOvary                                                                    1.566
## Tissue_CorrectedPlacenta                                                                -7.534
## Tissue_CorrectedSmall Intestine                                                         -1.128
## Tissue_CorrectedSpleen                                                                  -2.842
## Tissue_CorrectedStomach                                                                  3.994
## Tissue_CorrectedTail                                                                     1.742
## Tissue_CorrectedTestis                                                                  -5.505
## GenotypeAPC:min/+  ApclacIGY:ApclacIGY/+  Dnmt1:LO/LO VilLacIGY:+/+  Villin-mKate2:+/+      NA
## `CGI-mean`                                                                               9.698
## `CTCF-mean`                                                                             -2.934
## `Clock-mean`                                                                             5.018
## `SNP-mean`                                                                               1.051
## `SpermMeth-mean`                                                                         7.081
## `VMR-mean`                                                                              -3.226
##                                                                                        Pr(&gt;|t|)
## (Intercept)                                                                            0.021114
## SexMale                                                                                0.000106
## Strain_Corrected129-C57 Mixture                                                        0.434871
## Strain_Corrected129S1/SvlmJ                                                            0.010981
## Strain_CorrectedA/J                                                                    0.004694
## Strain_CorrectedAKR/J                                                                  0.067588
## Strain_CorrectedAthymic_Nude                                                           0.072865
## Strain_CorrectedBALB/cJ                                                                0.014354
## Strain_CorrectedBTBR_T+ ITpr3tf/J                                                      0.004193
## Strain_CorrectedC3H/HeJ                                                                0.014694
## Strain_CorrectedC57BL/10J                                                              0.000776
## Strain_CorrectedC57BL/6J                                                               0.001114
## Strain_CorrectedC57BL/6NJ                                                              0.000252
## Strain_CorrectedC57L/J                                                                 0.135545
## Strain_CorrectedC58/J                                                                  0.000134
## Strain_CorrectedCAST/EiJ                                                               0.017405
## Strain_CorrectedCBA/J                                                                  0.002255
## Strain_CorrectedDBA/1J                                                                 0.022123
## Strain_CorrectedDBA/2J                                                                 0.006022
## Strain_CorrectedFVB/NJ                                                                 0.001248
## Strain_CorrectedKK/HiJ                                                                 0.000708
## Strain_CorrectedLP/J                                                                   0.013701
## Strain_CorrectedMixed                                                                  0.099587
## Strain_CorrectedMOLF/EiJ                                                               5.61e-05
## Strain_CorrectedNOD/ShiLtJ                                                             0.000725
## Strain_CorrectedNSG                                                                    0.113889
## Strain_CorrectedNZB/BINJ                                                               0.004873
## Strain_CorrectedNZO/HlLtJ                                                              0.001194
## Strain_CorrectedNZW/LacJ                                                               0.001054
## Strain_CorrectedPWK/PhJ                                                                7.65e-05
## Strain_CorrectedWSB/EiJ                                                                0.095192
## Tissue_CorrectedCecum                                                                  0.246081
## Tissue_CorrectedEsophagus                                                              0.403407
## Tissue_CorrectedFat                                                                    0.059531
## Tissue_CorrectedFetal Brain                                                            0.240350
## Tissue_CorrectedFetal Intestine                                                        0.211825
## Tissue_CorrectedFetal Limb                                                             0.363440
## Tissue_CorrectedFrontal Lobe Brain                                                     0.015367
## Tissue_CorrectedHeart                                                                  0.035998
## Tissue_CorrectedHind Brain                                                             0.001404
## Tissue_CorrectedKidney                                                                 0.000598
## Tissue_CorrectedLiver                                                                  0.488633
## Tissue_CorrectedLung                                                                   0.007087
## Tissue_CorrectedMuscle                                                                 0.058985
## Tissue_CorrectedOvary                                                                  0.118112
## Tissue_CorrectedPlacenta                                                               3.19e-13
## Tissue_CorrectedSmall Intestine                                                        0.259853
## Tissue_CorrectedSpleen                                                                 0.004705
## Tissue_CorrectedStomach                                                                7.71e-05
## Tissue_CorrectedTail                                                                   0.082215
## Tissue_CorrectedTestis                                                                 6.54e-08
## GenotypeAPC:min/+  ApclacIGY:ApclacIGY/+  Dnmt1:LO/LO VilLacIGY:+/+  Villin-mKate2:+/+       NA
## `CGI-mean`                                                                              &lt; 2e-16
## `CTCF-mean`                                                                            0.003530
## `Clock-mean`                                                                           7.80e-07
## `SNP-mean`                                                                             0.293664
## `SpermMeth-mean`                                                                       6.27e-12
## `VMR-mean`                                                                             0.001357
##                                                                                           
## (Intercept)                                                                            *  
## SexMale                                                                                ***
## Strain_Corrected129-C57 Mixture                                                           
## Strain_Corrected129S1/SvlmJ                                                            *  
## Strain_CorrectedA/J                                                                    ** 
## Strain_CorrectedAKR/J                                                                  .  
## Strain_CorrectedAthymic_Nude                                                           .  
## Strain_CorrectedBALB/cJ                                                                *  
## Strain_CorrectedBTBR_T+ ITpr3tf/J                                                      ** 
## Strain_CorrectedC3H/HeJ                                                                *  
## Strain_CorrectedC57BL/10J                                                              ***
## Strain_CorrectedC57BL/6J                                                               ** 
## Strain_CorrectedC57BL/6NJ                                                              ***
## Strain_CorrectedC57L/J                                                                    
## Strain_CorrectedC58/J                                                                  ***
## Strain_CorrectedCAST/EiJ                                                               *  
## Strain_CorrectedCBA/J                                                                  ** 
## Strain_CorrectedDBA/1J                                                                 *  
## Strain_CorrectedDBA/2J                                                                 ** 
## Strain_CorrectedFVB/NJ                                                                 ** 
## Strain_CorrectedKK/HiJ                                                                 ***
## Strain_CorrectedLP/J                                                                   *  
## Strain_CorrectedMixed                                                                  .  
## Strain_CorrectedMOLF/EiJ                                                               ***
## Strain_CorrectedNOD/ShiLtJ                                                             ***
## Strain_CorrectedNSG                                                                       
## Strain_CorrectedNZB/BINJ                                                               ** 
## Strain_CorrectedNZO/HlLtJ                                                              ** 
## Strain_CorrectedNZW/LacJ                                                               ** 
## Strain_CorrectedPWK/PhJ                                                                ***
## Strain_CorrectedWSB/EiJ                                                                .  
## Tissue_CorrectedCecum                                                                     
## Tissue_CorrectedEsophagus                                                                 
## Tissue_CorrectedFat                                                                    .  
## Tissue_CorrectedFetal Brain                                                               
## Tissue_CorrectedFetal Intestine                                                           
## Tissue_CorrectedFetal Limb                                                                
## Tissue_CorrectedFrontal Lobe Brain                                                     *  
## Tissue_CorrectedHeart                                                                  *  
## Tissue_CorrectedHind Brain                                                             ** 
## Tissue_CorrectedKidney                                                                 ***
## Tissue_CorrectedLiver                                                                     
## Tissue_CorrectedLung                                                                   ** 
## Tissue_CorrectedMuscle                                                                 .  
## Tissue_CorrectedOvary                                                                     
## Tissue_CorrectedPlacenta                                                               ***
## Tissue_CorrectedSmall Intestine                                                           
## Tissue_CorrectedSpleen                                                                 ** 
## Tissue_CorrectedStomach                                                                ***
## Tissue_CorrectedTail                                                                   .  
## Tissue_CorrectedTestis                                                                 ***
## GenotypeAPC:min/+  ApclacIGY:ApclacIGY/+  Dnmt1:LO/LO VilLacIGY:+/+  Villin-mKate2:+/+    
## `CGI-mean`                                                                             ***
## `CTCF-mean`                                                                            ** 
## `Clock-mean`                                                                           ***
## `SNP-mean`                                                                                
## `SpermMeth-mean`                                                                       ***
## `VMR-mean`                                                                             ** 
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Residual standard error: 134.9 on 409 degrees of freedom
##   (1 observation deleted due to missingness)
## Multiple R-squared:  0.6907, Adjusted R-squared:  0.6484 
## F-statistic: 16.31 on 56 and 409 DF,  p-value: &lt; 2.2e-16</code></pre>
<p>```</p>
</div>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
    // active menu
    var href = window.location.pathname;
    href = href.substr(href.lastIndexOf('/') + 1);
    if (href === '') href = './';
    $('a[href="' + href + '"]').parent().addClass('active');
});
</script>
<script>
// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>


</body>
</html>
